spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindImproperSynchronization.java

```diff
@@ -0,0 +1,807 @@
+    }
+
+    @Override
+    public void visit(Method obj) {
+        try {
+            analyzeAssignments(obj, getClassContext());
+        } catch (CFGBuilderException e) {
+            AnalysisContext.logError("Error building CFG for " + obj, e);
+        }
+
+        if (isInitializerMethod(obj.getName())) {
+            return;
+        }
+
+        XMethod xMethod = getXMethod();
+        if (xMethod.isPublic() && xMethod.isSynchronized()) {
+            if (xMethod.isStatic()) {
+                bugReporter.reportBug(new BugInstance(
+                        this, "US_UNSAFE_STATIC_METHOD_SYNCHRONIZATION", NORMAL_PRIORITY)
+                        .addClassAndMethod(this));
+            }
+            synchronizedMethods.add(xMethod);
+        }
+
+        SignatureParser signature = new SignatureParser(xMethod.getSignature());
+        if ("readResolve".equals(obj.getName()) && "()Ljava/lang/Object;".equals(obj.getSignature())
+                && Subtypes2.instanceOf(getThisClass(), "java.io.Serializable")) {
+            return;
+        }
+
+        if (!(xMethod.isStatic() || xMethod.isPublic())) {
+            return;
+        }
+
+        String returnType = signature.getReturnTypeSignature();
+        String javaStyleClassType = "L" + getClassName() + ";";
+        if (returnType.equals(javaStyleClassType)) {
+            if ("clone".equals(obj.getName())) {
+                return;
+            }
+            exposingMethods.add(xMethod);
+        } else {
+            String sourceSig = xMethod.getSourceSignature();
+            if (sourceSig != null) {
+                String typePattern = "^.*<.*" + javaStyleClassType + ".*>;$";
+                GenericSignatureParser genericSignature = new GenericSignatureParser(sourceSig);
+                String genericReturnValue = genericSignature.getReturnTypeSignature();
+                if (genericReturnValue.matches(typePattern)) {
+                    exposingMethods.add(xMethod);
+                }
+            }
+        }
+
+    }
+
+    @Override
+    public void visit(Field obj) {
+        XField xField = getXField();
+        if (isCollection(xField)) {
+            try {
+                if (isInherited(xField)) {
+                    if (!xField.isPrivate()) {
+                        badBackingCollections.add(xField);
+                    }
+
+                    findExposureInHierarchy(xField)
+                            .forEach(method -> collectionAccessorsInHierarchy.add(xField, method));
+                } else {
+                    if (xField.isPublic()) {
+                        badBackingCollections.add(xField);
+                    }
+
+                    if (xField.isProtected() && !getThisClass().isFinal()) {
+                        inheritableBackingCollections.add(xField);
+                    } else if (isPackagePrivate(xField)) {
+                        inheritableBackingCollections.add(xField);
+                    }
```

Comment:
As far as I can see the body of these two `if`s are the same. Can you please combine them?

Reply:
I combined them, thank you!